artsy/hokusai
# feat: switch macos pyinstaller build to use onedir

To resolve https://github.com/artsy/hokusai/issues/296

The start-up lag mentioned in that issue is due to our [bundling dependencies into a single executable](https://pyinstaller.readthedocs.io/en/stable/operating-mode.html#how-the-one-folder-program-works
), which is known as PyInstaller One File mode. In this mode, every time Hokusai is run, the PyInstaller bootloader decompresses the one-file bundle into a temporary folder, hence the lag.

Proposal: Switch the MacOS build to PyInstaller [One Folder](https://pyinstaller.readthedocs.io/en/stable/operating-mode.html#how-the-one-folder-program-works) (a.k.a. onedir) mode which does not involve decompression at runtime.

Testing onedir-mode-built Hokusai on my local:
```
artsy:hokusai-sandbox jxu$ time hokusai version
1.0.1

real    0m2.824s
user    0m0.396s
sys     0m0.125s
artsy:hokusai-sandbox jxu$ time hokusai version
1.0.1

real    0m0.454s
user    0m0.359s
sys     0m0.092s
```

The lag on first invocation is same (bad) as before, but in subsequent invocations, the lag is much less.

This PR does not change the Linux build, because the lag on Linux seems tolerable, so we continue to enjoy easy (one-file) installation on Linux.

The changes are:
- Switch MacOS build to onedir mode
- Update Homebrew formula to install onedir build
- Introduce Homebrew formula for installing Beta. Currently, beta can be installed on MacOS by a simple [Bash script](https://github.com/artsy/hokusai#distributing-hokusai). That script will no longer work for onedir build.
